modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/data_renderers/ActionsDropdownRenderer.js

```diff
@@ -359,37 +384,55 @@ function ActionsDropdownRenderer({actions, itemData, itemId}) {
 		return <ClayLoadingIndicator small />;
 	}
 
-	const renderItems = (items) =>
-		items.map(({items: nestedItems = [], separator, type, ...item}, i) => {
-			if (type === 'group') {
-				return (
-					<ClayDropDown.Group {...item}>
-						{separator && <ClayDropDown.Divider />}
+	const renderItems = ({actions, quickActionItems = false}) =>
+		actions.map(
+			({items: nestedItems = [], separator, type, ...item}, i) => {
+				if (type === 'group') {
+					return (
+						<ClayDropDown.Group {...item}>
+							{separator && <ClayDropDown.Divider />}
+
+							{renderItems({
+								actions: nestedItems,
+								quickActionItems,
+							})}
+						</ClayDropDown.Group>
+					);
+				}
 
-						{renderItems(nestedItems)}
-					</ClayDropDown.Group>
+				return (
+					<ActionItem
+						action={item}
+						closeMenu={() => setMenuActive(false)}
+						handleAction={handleAction}
+						itemData={itemData}
+						itemId={itemId}
+						key={i}
+						method={item.method ?? item.data?.method}
+						quickActionItem={quickActionItems}
```

Comment:
continued from https://github.com/liferay-frontend/liferay-portal/pull/2222#discussion_r875050663:
>Between quickActionItem and actionItem the only thing changes is the markup, because we do not want to render always a dropdown item component, although it does the job...
They both share the logic for the event handler, that's why I keep them together

I see your point, but I think we sacrifice too much readability for DRY. Especially this line is hard to understand on a glance:
```js
quickActionItem={quickActionItems}
```
I think the code would be much easier to read if we had a couple of internal components, that contain specific markup and share handler as a prop. Something like
```js
<DropdownActionItem
    onClick={handleClick}
    ...
>
<DropdownActionItem>
```
and
```js
<QuickActionItem
    onClick={handleClick}
    ...
>
<QuickActionItem>
```
Can we do this without too much duplication?

Reply:
I see your point but I'm not sure... I can try to declare and share the handler from the parent component, this way I can avoid that "dirty" boolean I'm using. I agree it's not super nice 😅 
The downside of this approach is that `renderItems` function will become more complex. i give it a try and let you know

Reply:
Still not sure about this @markocikos ...

If I move up the handlers to the parent function I does not much sense to keep the `ActionItem` component, which at the end leads to a large and more complex `renderItems` map, also It does not make much sense to render the dropdown/quick action items as components because at the end they will only output a really tiny markup...

Also, I feel much safer to pass down a boolean than move up that handlers.

What about a simple renaming:

```
isQuickActionItem={renderQuickActionItems}
```

Hope I explained myself. Let me know, thxs!

Reply:
I see your point, and if we needed the quickest solution, that would be it. But, this epic seems to me like a great opportunity to improve architecture of action components. Here are my thoughts on broader structure:
- Quick actions should not even be in `ActionsDropdownRenderer`. Dropdown is a component next to quick actions, that has it's own specifics, like logic when the dropdown is open or closed. A more cleaner structure would be to have `QuickActionsMenu` component, with `QuickAction` internal components. We could extract any common logic for action definitions to utilities (`/actions/doAction(action)`?), a new top component (`ActionsMenu`?) or a new context (`ActionsContext`?).
- We don't need or want to reuse `renderItems` for quick actions. It contains logic that is specific to to dropdown, such as `ClayDropDown.Group` markup.
- `ActionsDropdownRenderer` should not be under `data_renderers`. It is not a way to render data. It should be called `ActionsDropdown`, maybe under `actions` dir? I do realize that current `actions` dir has nothing to do with item actions, but I'm not sure where it would fit more.
- `ActionsDropdownRenderer` and `ActionLinkRenderer` have a lot of duplicated code, especially when it comes to action targets handling. There is some divergence, but I bet this is only because fixes we made on one and not another, and people forgot to update the other. If we extract event handling to utils, we could remove this duplication.

What do you think?